File: examples/servers/dictionary/gateway_setup/setup-gateway.ts
#!/usr/bin/env node

import { readFileSync } from "fs";
import { STSClient, GetCallerIdentityCommand } from "@aws-sdk/client-sts";
import {
  CloudFormationClient,
  DescribeStacksCommand,
} from "@aws-sdk/client-cloudformation";
import { SSMClient, PutParameterCommand } from "@aws-sdk/client-ssm";
import {
  BedrockAgentCoreControlClient,
  CreateGatewayCommand,
  CreateGatewayTargetCommand,
  GetGatewayCommand,
} from "@aws-sdk/client-bedrock-agentcore-control";

async function main() {
  const region = "us-west-2";

  // Get unique suffix for integration tests
  const suffix =
    "INTEG_TEST_ID" in process.env ? `-${process.env["INTEG_TEST_ID"]}` : "";

  // Get account ID
  const stsClient = new STSClient({ region });
  const identity = await stsClient.send(new GetCallerIdentityCommand({}));
  const accountId = identity.Account!;

  // Load tools list
  const toolsConfig = JSON.parse(readFileSync("tools-list.json", "utf8"));

  // Get Lambda function ARN from CloudFormation stack
  const cfClient = new CloudFormationClient({ region });
  const stackResponse = await cfClient.send(
    new DescribeStacksCommand({
      StackName: `LambdaMcpServer-Dictionary${suffix}`,
    })
  );

  let lambdaFunctionArn: string | undefined;
  for (const output of stackResponse.Stacks![0].Outputs || []) {
    if (output.OutputKey?.includes("ServerFunctionOutput")) {
      lambdaFunctionArn = output.OutputValue;
      break;
    }
  }

  if (!lambdaFunctionArn) {
    throw new Error("Lambda function ARN not found in stack outputs");
  }

  // Get Cognito authorizer info from Auth stack
  const authStackResponse = await cfClient.send(
    new DescribeStacksCommand({
      StackName: "LambdaMcpServer-Auth",
    })
  );

  const outputs: Record<string, string> = {};
  for (const output of authStackResponse.Stacks![0].Outputs || []) {
    outputs[output.OutputKey!] = output.OutputValue!;
  }

  const authorizerConfig = {
    customJWTAuthorizer: {
      allowedClients: [
        outputs.InteractiveOAuthClientId,
        outputs.AutomatedOAuthClientId,
      ],
      discoveryUrl: `${outputs.IssuerDomain}/.well-known/openid-configuration`,
    },
  };

  // Create Gateway
  const agentCoreClient = new BedrockAgentCoreControlClient({
    region,
    maxAttempts: 10,
  });
  const roleArn = `arn:aws:iam::${accountId}:role/mcp-lambda-example-agentcore-gateways`;

  let gatewayName = `LambdaMcpServer-Dictionary-Gateway${suffix}`;
  if (gatewayName.length > 50) {
    gatewayName = gatewayName.substring(0, 50).replace(/-+$/, '');
  }

  const gateway = await agentCoreClient.send(
    new CreateGatewayCommand({
      name: gatewayName,
      roleArn,
      protocolType: "MCP",
      authorizerType: "CUSTOM_JWT",
      authorizerConfiguration: authorizerConfig,
      exceptionLevel: "DEBUG",
    })
  );

  // Wait for gateway to be ready
  let attempts = 0;
  let gatewayReady = false;
  while (!gatewayReady && attempts < 30) {
    const gatewayStatus = await agentCoreClient.send(
      new GetGatewayCommand({
        gatewayIdentifier: gateway.gatewayId!,
      })
    );
    if (gatewayStatus.status === "READY") {
      gatewayReady = true;
    } else {
      await new Promise((resolve) => setTimeout(resolve, 2000));
      attempts++;
    }
  }

  if (!gatewayReady) {
    throw new Error("Gateway did not become ready within 30 attempts");
  }

  // Save Gateway URL and ID to SSM
  const ssmClient = new SSMClient({ region });
  const gatewayInfo = {
    gatewayId: gateway.gatewayId,
    url: gateway.gatewayUrl,
  };

  await ssmClient.send(
    new PutParameterCommand({
      Name: `LambdaMcpServer-Dictionary-Gateway${suffix}`,
      Value: JSON.stringify(gatewayInfo),
      Type: "String",
      Overwrite: true,
    })
  );

  // Create Lambda target
  const targetPayload = {
    lambdaArn: lambdaFunctionArn,
    toolSchema: { inlinePayload: toolsConfig.tools },
  };

  await agentCoreClient.send(
    new CreateGatewayTargetCommand({
      gatewayIdentifier: gateway.gatewayId,
      name: "dictionary-target",
      targetConfiguration: { mcp: { lambda: targetPayload } },
      credentialProviderConfigurations: [
        { credentialProviderType: "GATEWAY_IAM_ROLE" },
      ],
    })
  );

  console.log("Gateway created successfully:");
  console.log(`  ID: ${gateway.gatewayId}`);
  console.log(`  URL: ${gateway.gatewayUrl}`);
  console.log(`  Lambda ARN: ${lambdaFunctionArn}`);
  console.log(
    `  Saved to SSM parameter: LambdaMcpServer-Dictionary-Gateway${suffix}`
  );
}

main().catch(console.error);

File: examples/servers/book-search/gateway_setup/setup_gateway.py
#!/usr/bin/env python3

import json
import os
import boto3
from botocore.config import Config
from bedrock_agentcore_starter_toolkit.operations.gateway.client import GatewayClient


def main():
    # Get unique suffix for integration tests
    suffix = f'-{os.environ["INTEG_TEST_ID"]}' if "INTEG_TEST_ID" in os.environ else ""

    # Get account ID
    sts = boto3.client("sts")
    account_id = sts.get_caller_identity()["Account"]

    # Load tools list
    with open("tools-list.json", "r") as f:
        tools_config = json.load(f)

    # Get Lambda function ARN from CloudFormation stack
    cf_client = boto3.client("cloudformation", region_name="us-west-2")
    stack_response = cf_client.describe_stacks(
        StackName=f"LambdaMcpServer-BookSearch{suffix}"
    )

    lambda_function_arn = None
    for output in stack_response["Stacks"][0]["Outputs"]:
        if "ServerFunction" in output["OutputKey"]:
            lambda_function_arn = output["OutputValue"]
            break

    if not lambda_function_arn:
        raise ValueError("Lambda function ARN not found in stack outputs")

    # Get Cognito authorizer info from Auth stack
    auth_stack_response = cf_client.describe_stacks(StackName="LambdaMcpServer-Auth")

    outputs = {
        output["OutputKey"]: output["OutputValue"]
        for output in auth_stack_response["Stacks"][0]["Outputs"]
    }

    authorizer_config = {
        "customJWTAuthorizer": {
            "allowedClients": [
                outputs["InteractiveOAuthClientId"],
                outputs["AutomatedOAuthClientId"],
            ],
            "discoveryUrl": f"{outputs['IssuerDomain']}/.well-known/openid-configuration",
        }
    }

    # Create Gateway client for us-west-2
    gateway_client = GatewayClient(region_name="us-west-2")
    retry_config = Config(
        retries={
            "max_attempts": 10,
            "mode": "standard",
        }
    )
    boto_client = boto3.client(
        "bedrock-agentcore-control", region_name="us-west-2", config=retry_config
    )
    gateway_client.client = boto_client

    # Create Gateway
    role_arn = f"arn:aws:iam::{account_id}:role/mcp-lambda-example-agentcore-gateways"

    gateway_name = f"LambdaMcpServer-BookSearch-Gateway{suffix}"
    if len(gateway_name) > 48:
        gateway_name = gateway_name[:48].rstrip('-')

    gateway = gateway_client.create_mcp_gateway(
        name=gateway_name,
        role_arn=role_arn,
        authorizer_config=authorizer_config,
        enable_semantic_search=False,
    )

    # Save Gateway URL and ID to SSM
    ssm_client = boto3.client("ssm", region_name="us-west-2")

    gateway_info = {
        "gatewayId": gateway["gatewayId"],
        "url": gateway["gatewayUrl"],
    }

    ssm_client.put_parameter(
        Name=f"LambdaMcpServer-BookSearch-Gateway{suffix}",
        Value=json.dumps(gateway_info),
        Type="String",
        Overwrite=True,
    )

    # Create Lambda target
    target_payload = {
        "lambdaArn": lambda_function_arn,
        "toolSchema": {"inlinePayload": tools_config["tools"]},
    }

    gateway_client.create_mcp_gateway_target(
        gateway=gateway,
        name="book-search-target",
        target_type="lambda",
        target_payload=target_payload,
    )

    print(f"Gateway created successfully:")
    print(f"  ID: {gateway['gatewayId']}")
    print(f"  URL: {gateway['gatewayUrl']}")
    print(f"  Lambda ARN: {lambda_function_arn}")
    print(f"  Saved to SSM parameter: LambdaMcpServer-BookSearch-Gateway{suffix}")


if __name__ == "__main__":
    main()
